port/osx/omrosbacktrace_impl.c

```diff
@@ -0,0 +1,413 @@
+	 * See https://opensource.apple.com/source/libunwind/libunwind-35.4/src/unw_getcontext.s.auto.html
+	 */
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_BP, &name, &reg);
+	unw_set_reg(cursor, UNW_X86_64_RBP, *(unw_word_t *)reg);
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_SP, &name, &reg);
+	/* Store call-site sp (skip return address) into RSP. */
+	unw_set_reg(cursor, UNW_X86_64_RSP, (*(unw_word_t *)reg) + 8);
+	infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_PC, &name, &reg);
+	ip = *(unw_word_t *)reg;
+
+	if (nAddr < (size - 1)) {
+		/* libunwind failed to iterate past _sigtramp because of a call to invalid address or a signal
+		 * received in kernel (right after syscall). In either case, the return address will be
+		 * pointed to by rsp: in the first case, return address pushed to stack right before call,
+		 * in the second case, syscall wrapper's caller pushed its return address. Record the ip and
+		 * skip the frame.
+		 */
+		array[nAddr] = (void *)ip;
+		nAddr += 1;
+		infoForControl(NULL, &info, OMRPORT_SIG_CONTROL_SP, &name, &reg);
+		ip = **(unw_word_t **)reg;
+
+		/* Record the return address. */
+		array[nAddr] = (void *)ip;
+		nAddr += 1;
+		unw_set_reg(cursor, UNW_REG_IP, ip);
+
+		while ((unw_step(cursor) > 0) && (nAddr < size)) {
+			unw_get_reg(cursor, UNW_REG_IP, &ip);
+			array[nAddr] = (void *)ip;
+			nAddr += 1;
+		}
+	}
+#endif /* defined(OMR_ARCH_X86) */
+
+	return nAddr;
+}
+
+/**
+ * Obtains the stack trace for the current thread using libunwind and stores it in array. libunwind is used
+ * instead of libc backtrace as it handles signal handler frames correctly, whereas backtrace may return an
+ * incorrect frame for the frame before the signal handler.
+ *
+ * @param[out] array the array to hold the frame addresses
+ * @param[in] size capacity of the array
+ *
+ * @return the number of addresses recorded in the array
+ */
+static uintptr_t
+unw_backtrace(void **array, const uintptr_t size)
+{
+	unw_cursor_t cursor;
+	unw_cursor_t prev;
+	unw_context_t uc;
+	unw_word_t ip = 0;
+	char buf[64];
+	uintptr_t nAddr = 0;
+
+	unw_getcontext(&uc);
+	unw_init_local(&cursor, &uc);
+	while ((unw_step(&cursor) > 0) && (nAddr < size)) {
+		unw_get_reg(&cursor, UNW_REG_IP, &ip);
+		array[nAddr] = (void *)ip;
+		nAddr += 1;
+
+		/* Save frame for check below. */
+		memcpy(&prev, &cursor, sizeof(unw_cursor_t));
+	}
+
+	/* Darwin's libunwind cannot handle sigtramp in some cases. If the final frame iterated is sigtramp,
+	 * there are more frames but need to manually restore registers.
+	 */
+	if (0 == unw_get_proc_name(&prev, buf, sizeof(buf), &ip)) {
+		if (0 == strcmp(buf, "_sigtramp")) {
+			nAddr = handle_sigtramp(&prev, array, size, nAddr);
+		}
+	}
+
+	return nAddr;
+}
+
+/**
+ * NULL handler. We only care about preventing the signal from propagating up the call stack, no need to do
+ * anything in the handler.
+ */
+static uintptr_t
+handler(OMRPortLibrary *portLibrary, uint32_t gpType, void *gpInfo, void *userData)
```

Comment:
Missing function parameters' and return type description: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-function-parameters-clearly